src/group_theory/conjugates.lean

```diff
@@ -0,0 +1,142 @@
+/-
+
+Author: Michael Howes and Scott Morrison, 2019
+
+The definition of conjugates in a group and the definition
+of the normal closure of a subset of a group.
+-/
+import algebra.group
+import group_theory.subgroup
+import tactic.rewrite
+
+variable {α : Type}
+variable [group α]
+
+
+
+/-- Two elements a b of a group are conjugates if there exists a
+c such that c*a*c⁻¹ = b -/
+def is_conjugate (a b : α) : Prop := ∃ c : α, c*a*c⁻¹ = b
+
+/-- Some basic properties of conjugates -/
+@[simp] lemma conj_inv {a b : α} : (b*a*b⁻¹)⁻¹ = b*a⁻¹*b⁻¹ :=
+begin
+  rw [mul_inv_rev _ b⁻¹, mul_inv_rev b _, inv_inv, mul_assoc],
+end
+
+@[simp] lemma conj_mul {a b c : α} : (b*a*b⁻¹)*(b*c*b⁻¹) = b*(a*c)*b⁻¹ :=
+begin
+  assoc_rw inv_mul_cancel_right,
+  repeat {rw mul_assoc},
+end
+
+namespace is_conjugate
+/-- is_conjugate is an equivalence relation -/
+lemma refl {a : α} : is_conjugate a a := exists.intro (1:α) (by simp)
+
+lemma symm_aux {a b c : α} : c*a*c⁻¹ = b → a=c⁻¹*b*c  :=
+λ h, by rw [mul_assoc, ←(eq_inv_mul_of_mul_eq (eq_mul_of_mul_inv_eq h))]
+
+lemma symm {a b : α} : is_conjugate a b → is_conjugate b a :=
+λ ⟨c,h⟩, ⟨c⁻¹, by simpa only [inv_inv] using eq.symm (symm_aux h)⟩
+
+lemma trans_aux {a c d e : α} : d * a * d⁻¹ = e⁻¹ * c * e → e*d*a*(d⁻¹*e⁻¹) = c :=
+λ h,
+begin
+  assoc_rw h,
+  simp,
+end
+
+lemma trans {a b c : α} : is_conjugate a b → is_conjugate b c → is_conjugate a c :=
+λ ⟨d,w⟩ ⟨e,h⟩,
+begin
+  rw (symm_aux h) at w,
+  exact ⟨e*d, by simpa only [mul_inv_rev] using (trans_aux w)⟩
+end
+
+end is_conjugate
+
+/-- Here the normal closure of a set is defined as the subgroup closure of the
+set of conjugates of the given conjugates -/
+
+variables {s : set α}
+
+/-- Given an element a, conjugates a is the set of conjugates -/
+def conjugates (a : α) : set α := is_conjugate a
+
+lemma self_mem_conjugates {a : α} : a ∈ conjugates a := is_conjugate.refl
+
+/-- Given a set s, conjugates_of_set s is the set of all conjugates of
+the elements of s. -/
+def conjugates_of_set (s : set α) : set α :=
+  λ a : α, ∃ x ∈ s, a ∈ conjugates x
+
+theorem mem_conjugates {x : α} : x ∈ s → x ∈ conjugates_of_set s :=
+  λ h, Exists.intro x (Exists.intro h self_mem_conjugates)
+
```

Comment:
It would be nice to add a lemma `mem_conjugates_of_set_iff` around here. It can be useful as rewriting lemma. (The lemma says that `x \in conjugates_of_set s` if and only if there exists an `a` such that..., and is currently true by definition. But if you follow Mario's advise upstairs, this lemma will be useful.)

Reply:
I've added this. I'm not sure how useful it is as I didn't end up using Mario's suggested definition.